server/src/main/java/org/opensearch/common/throttle/AdmissionController.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.throttle;
+
+/**
+ * Aims to provide resource based request admission control.
+ * Provides methods for any tracking-object that can be incremented (such as memory size),
+ * Admission control can be applied if configured limit has been reached.
+ */
+public interface AdmissionController {
+
+    /**
+     * @return name of the admission-controller
+     */
+    String getName();
+
+    /**
+     * @return admission-controller enabled status
+     */
+    boolean isEnabled();
+
+    /**
+     * Increment the tracking-object with provided value.
+     * Apply the admission control if threshold is breached.
+     * Mostly applicable while acquiring the quota.
+     *
+     * @param count value to incrementation the resource racking-object with.
+     * @return count/value acquired from the resource tracking object.
+     */
+    long acquire(long count);
+
+    /**
+     * Decrement the tracking-object with provided value and do not apply the admission control.
+     * Mostly applicable while remitting the quota.
+     *
+     * @param count value to decrement the resource tracking-object with.
+     * @return count/value released to the resource tracking object.
+     */
+    long release(long count);
+
```

Comment:
These are internal working of an Admission controller, so doesn't add value exposing them

Reply:
Hi @Bukhtawar,
Thanks for reviewing pull request.

I was planning to create `AdmissionControlHandler`, a netty handler that extends `ChannelDuplexHandler`.

* `acquire` function would be used from `channelRead` function of `AdmissionControlHandler`
* `release` function would be used from `write` function of `AdmissionControlHandler`

Let me know your thoughts about it.

Reply:
Let me rephrase it on why I call them internal working
For a particular request, release cannot release more than what it acquired. Recommend having a method addBytesAndMaybeBreak

Reply:
Thanks @Bukhtawar. Instead of separate function for `acquire` and `release`, I've added  `addBytesAndMaybeBreak`  function which returns `Releasable` which would be used to release tokens which were allocated to a particular request.